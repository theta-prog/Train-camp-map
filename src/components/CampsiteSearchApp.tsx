'use client'

import CampsiteList from '@/components/CampsiteList'
import LanguageSwitcher from '@/components/LanguageSwitcher'
import SearchFilters from '@/components/SearchFilters'
import { Campsite } from '@/types/campsite'
import { extractPriceFromString, fetchCampsites } from '@/utils/campsiteApi'
import { APIProvider } from '@vis.gl/react-google-maps'
import { useTranslations } from 'next-intl'
import dynamic from 'next/dynamic'
import { useCallback, useEffect, useState } from 'react'

// MapComponentを動的インポートしてSSRを無効化
const MapComponent = dynamic(
  /* istanbul ignore next */ () => import('@/components/MapComponent'),
  {
    ssr: false,
    loading: () => (
      <div className="flex items-center justify-center" style={{ width: '100%', height: '600px' }}>
        <div className="text-center">
          <div className="inline-block animate-spin rounded-full h-8 w-8 border-b-2 border-green-600"></div>
          <p className="mt-2 text-gray-600">Loading map...</p>
        </div>
      </div>
    )
  }
)

export default function CampsiteSearchApp() {
  const t = useTranslations()
  const [selectedCampsite, setSelectedCampsite] = useState<Campsite | null>(null)
  const [allCampsites, setAllCampsites] = useState<Campsite[]>([])
  const [filteredCampsites, setFilteredCampsites] = useState<Campsite[]>([])
  const [isLoading, setIsLoading] = useState(true)
  const [error, setError] = useState<string | null>(null)
  const [searchFilters, setSearchFilters] = useState({
    keyword: '',
    maxPrice: 10000,
    facilities: [] as string[],
    activities: [] as string[]
  })
  const [isFiltersInitialized, setIsFiltersInitialized] = useState(false)

  const apiKey = process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY

  // キャンプサイトデータを取得
  useEffect(() => {
    async function loadCampsites() {
      try {
        setIsLoading(true)
        setError(null)
        const campsites = await fetchCampsites()
        setAllCampsites(campsites)
        setFilteredCampsites(campsites)
      } catch (err) {
        console.error('Failed to load campsites:', err)
        setError('キャンプサイトの読み込みに失敗しました')
      } finally {
        setIsLoading(false)
      }
    }

    loadCampsites()
  }, [])

  const handleFilterChange = useCallback((filters: typeof searchFilters) => {
    setSearchFilters(filters)
    
    if (!isFiltersInitialized) {
      setIsFiltersInitialized(true)
      return // 初回は何もしない
    }
    
    let filtered = allCampsites.filter((campsite: Campsite) => {
      const keywordLower = filters.keyword.toLowerCase()
      const keywordMatch = !filters.keyword || 
        campsite.name.toLowerCase().includes(keywordLower) ||
        campsite.address.toLowerCase().includes(keywordLower) ||
        campsite.nearestStation.toLowerCase().includes(keywordLower)
      
      const priceValue = extractPriceFromString(campsite.price)
      const priceWithinLimit = priceValue <= filters.maxPrice
      
      const facilitiesMatch = filters.facilities.length === 0 || 
        filters.facilities.some((facility: string) => campsite.facilities.includes(facility))
      
      const activitiesMatch = filters.activities.length === 0 || 
        filters.activities.some((activity: string) => campsite.activities.includes(activity))
      
      return keywordMatch && priceWithinLimit && facilitiesMatch && activitiesMatch
    })
    
    setFilteredCampsites(filtered)
  }, [isFiltersInitialized, allCampsites])

  if (!apiKey) {
    return (
      <div className="min-h-screen bg-gray-50 flex items-center justify-center">
        <div className="text-center">
          <p className="text-gray-600 mb-2">{t('map.apiKeyMissing')}</p>
          <p className="text-sm text-gray-500">
            {t('map.apiKeyInstruction')}
          </p>
        </div>
      </div>
    )
  }

  if (isLoading) {
    return (
      <APIProvider apiKey={apiKey}>
        <div className="min-h-screen bg-gray-50 flex items-center justify-center">
          <div className="text-center">
            <div className="inline-block animate-spin rounded-full h-12 w-12 border-b-2 border-green-600"></div>
            <p className="mt-4 text-gray-600">キャンプサイトを読み込み中...</p>
          </div>
        </div>
      </APIProvider>
    )
  }

  if (error) {
    return (
      <APIProvider apiKey={apiKey}>
        <div className="min-h-screen bg-gray-50 flex items-center justify-center">
          <div className="text-center">
            <p className="text-red-600 mb-2">{error}</p>
            <button 
              onClick={() => window.location.reload()} 
              className="px-4 py-2 bg-green-600 text-white rounded hover:bg-green-700"
            >
              再読み込み
            </button>
          </div>
        </div>
      </APIProvider>
    )
  }

  return (
    <APIProvider apiKey={apiKey}>
      <div className="min-h-screen bg-gray-50">
        <header className="bg-green-600 text-white py-4">
          <div className="container mx-auto px-4">
            <div className="flex justify-between items-center">
              <div>
                <h1 className="text-2xl font-bold">
                  {t('title')}
                </h1>
                <p className="text-green-100">
                  {t('subtitle')}
                </p>
              </div>
              <LanguageSwitcher />
            </div>
          </div>
        </header>

        <div className="container mx-auto px-4 py-6">
          <div className="grid grid-cols-1 lg:grid-cols-4 gap-6">
            <div className="lg:col-span-1">
              <SearchFilters 
                onFilterChange={handleFilterChange}
                initialFilters={searchFilters}
              />
            </div>

            <div className="lg:col-span-2">
              <div className="bg-white rounded-lg shadow-md overflow-hidden">
                <MapComponent 
                  campsites={filteredCampsites}
                  onCampsiteSelect={setSelectedCampsite}
                  selectedCampsite={selectedCampsite}
                />
              </div>
            </div>

            <div className="lg:col-span-1">
              <CampsiteList 
                campsites={filteredCampsites}
                onCampsiteSelect={setSelectedCampsite}
                selectedCampsite={selectedCampsite}
              />
            </div>
          </div>
        </div>
      </div>
    </APIProvider>
  )
}
