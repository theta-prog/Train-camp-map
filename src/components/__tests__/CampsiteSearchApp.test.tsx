import '@testing-library/jest-dom'
import { render, screen, waitFor } from '@testing-library/react'
import CampsiteSearchApp from '../CampsiteSearchApp'

// サンプルデータ
const sampleCampsites = [
  {
    id: '1',
    name: { ja: '高尾山キャンプ場', en: 'Takao Mountain Campsite' },
    description: { ja: '高尾山の麓にある自然豊かなキャンプ場', en: 'A nature-rich campsite at the foot of Mount Takao' },
    location: { lat: 35.6422, lng: 139.2676 },
    address: { ja: '東京都八王子市高尾町', en: 'Takao, Hachioji City, Tokyo' },
    nearestStation: { ja: 'JR高尾駅', en: 'JR Takao Station' },
    accessTime: { ja: '徒歩15分', en: '15 min walk' },
    price: '¥2,000/泊',
    facilities: ['restroom', 'shower', 'parking'],
    activities: ['hiking', 'photography'],
    website: 'https://example.com',
    phone: '042-xxx-xxxx'
  }
]

// モック next-intl
jest.mock('next-intl', () => ({
  useTranslations: () => (key: string) => {
    const translations: Record<string, string> = {
      'title': 'Campsite Search',
      'subtitle': 'Find campsites accessible by train',
      'map.apiKeyMissing': 'Google Maps API key is missing',
      'map.apiKeyInstruction': 'Please set NEXT_PUBLIC_GOOGLE_MAPS_API_KEY',
    }
    return translations[key] || key
  }
}))

// API関数のモック
jest.mock('@/utils/campsiteApi', () => ({
  fetchCampsites: jest.fn(),
  extractPriceFromString: jest.fn((price: string) => {
    const match = price.match(/\d+/)
    return match ? parseInt(match[0]) : 0
  }),
  transformApiCampsiteToFrontend: jest.fn()
}))

// モック関数を取得
const { fetchCampsites } = require('@/utils/campsiteApi')
const mockFetchCampsites = fetchCampsites as jest.MockedFunction<typeof fetchCampsites>

// モック Google Maps API
jest.mock('@vis.gl/react-google-maps', () => ({
  APIProvider: ({ children }: any) => <div data-testid="api-provider">{children}</div>,
}))

// モック dynamic import
jest.mock('next/dynamic', () => {
  return function mockDynamic(_importFunc: any, options?: any) {
    const MockedComponent = (_props: any) => (
      <div data-testid="mock-map-component">
        {options?.loading && options.loading()}
      </div>
    )
    return MockedComponent
  }
})

// 子コンポーネントのモック
jest.mock('../SearchFilters', () => {
  return function MockSearchFilters({ onFilterChange }: any) {
    return (
      <div data-testid="mock-search-filters">
        <button 
          onClick={() => onFilterChange({ keyword: 'test', maxPrice: 5000, facilities: [], activities: [] })}
        >
          Change Filter
        </button>
      </div>
    )
  }
})

jest.mock('../CampsiteList', () => {
  return function MockCampsiteList({ campsites }: any) {
    return (
      <div data-testid="mock-campsite-list">
        {campsites.map((campsite: any) => (
          <div key={campsite.id}>
            {campsite.name.ja}
          </div>
        ))}
      </div>
    )
  }
})

jest.mock('../LanguageSwitcher', () => {
  return function MockLanguageSwitcher() {
    return <div data-testid="mock-language-switcher">Language Switcher</div>
  }
})

// 環境変数のモック
const originalEnv = process.env
beforeEach(() => {
  process.env = { ...originalEnv }
  mockFetchCampsites.mockResolvedValue(sampleCampsites)
})

afterEach(() => {
  process.env = originalEnv
  jest.clearAllMocks()
})

describe('CampsiteSearchApp', () => {
  describe('スナップショットテスト', () => {
    it('基本表示（APIキーあり）', async () => {
      process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY = 'test-api-key'
      
      const { container } = render(<CampsiteSearchApp locale="ja" />)
      
      // データが読み込まれるまで待つ
      await waitFor(() => {
        expect(screen.getByTestId('mock-campsite-list')).toBeInTheDocument()
      })
      
      expect(container.firstChild).toMatchSnapshot()
    })

    it('英語ロケール', async () => {
      process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY = 'test-api-key'
      
      const { container } = render(<CampsiteSearchApp locale="en" />)
      
      // データが読み込まれるまで待つ
      await waitFor(() => {
        expect(screen.getByTestId('mock-campsite-list')).toBeInTheDocument()
      })
      
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('基本機能', () => {
    it('APIキーがない場合エラーメッセージ表示', () => {
      delete process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY
      
      render(<CampsiteSearchApp locale="ja" />)
      
      expect(screen.getByText('Google Maps API key is missing')).toBeInTheDocument()
      expect(screen.getByText('Please set NEXT_PUBLIC_GOOGLE_MAPS_API_KEY')).toBeInTheDocument()
    })

    it('APIキーがある場合正常表示', async () => {
      process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY = 'test-api-key'
      
      render(<CampsiteSearchApp locale="ja" />)
      
      // データが読み込まれるまで待つ
      await waitFor(() => {
        expect(screen.getByText('Campsite Search')).toBeInTheDocument()
      })
      
      expect(screen.getByText('Find campsites accessible by train')).toBeInTheDocument()
      expect(screen.getByTestId('mock-search-filters')).toBeInTheDocument()
      expect(screen.getByTestId('mock-campsite-list')).toBeInTheDocument()
      expect(screen.getByTestId('mock-language-switcher')).toBeInTheDocument()
    })
  })
})
