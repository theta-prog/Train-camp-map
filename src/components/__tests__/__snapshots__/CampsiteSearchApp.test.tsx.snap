// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CampsiteSearchApp スナップショットテスト 基本表示（APIキーあり） 1`] = `
<div
  data-testid="api-provider"
>
  <div
    class="min-h-screen bg-gray-50"
  >
    <header
      class="bg-green-600 text-white py-4"
    >
      <div
        class="container mx-auto px-4"
      >
        <div
          class="flex justify-between items-center"
        >
          <div>
            <h1
              class="text-2xl font-bold"
            >
              Campsite Search
            </h1>
            <p
              class="text-green-100"
            >
              Find campsites accessible by train
            </p>
          </div>
          <div
            data-testid="mock-language-switcher"
          >
            Language Switcher
          </div>
        </div>
      </div>
    </header>
    <div
      class="container mx-auto px-4 py-6"
    >
      <div
        class="grid grid-cols-1 lg:grid-cols-4 gap-6"
      >
        <div
          class="lg:col-span-1"
        >
          <div
            data-testid="mock-search-filters"
          >
            <button>
              Change Filter
            </button>
          </div>
        </div>
        <div
          class="lg:col-span-2"
        >
          <div
            class="bg-white rounded-lg shadow-md overflow-hidden"
          >
            <div
              data-testid="mock-map-component"
            >
              <div
                class="flex items-center justify-center"
                style="width: 100%; height: 600px;"
              >
                <div
                  class="text-center"
                >
                  <div
                    class="inline-block animate-spin rounded-full h-8 w-8 border-b-2 border-green-600"
                  />
                  <p
                    class="mt-2 text-gray-600"
                  >
                    Loading map...
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="lg:col-span-1"
        >
          <div
            data-testid="mock-campsite-list"
          >
            <div>
              高尾山キャンプ場
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CampsiteSearchApp スナップショットテスト 英語ロケール 1`] = `
<div
  data-testid="api-provider"
>
  <div
    class="min-h-screen bg-gray-50"
  >
    <header
      class="bg-green-600 text-white py-4"
    >
      <div
        class="container mx-auto px-4"
      >
        <div
          class="flex justify-between items-center"
        >
          <div>
            <h1
              class="text-2xl font-bold"
            >
              Campsite Search
            </h1>
            <p
              class="text-green-100"
            >
              Find campsites accessible by train
            </p>
          </div>
          <div
            data-testid="mock-language-switcher"
          >
            Language Switcher
          </div>
        </div>
      </div>
    </header>
    <div
      class="container mx-auto px-4 py-6"
    >
      <div
        class="grid grid-cols-1 lg:grid-cols-4 gap-6"
      >
        <div
          class="lg:col-span-1"
        >
          <div
            data-testid="mock-search-filters"
          >
            <button>
              Change Filter
            </button>
          </div>
        </div>
        <div
          class="lg:col-span-2"
        >
          <div
            class="bg-white rounded-lg shadow-md overflow-hidden"
          >
            <div
              data-testid="mock-map-component"
            >
              <div
                class="flex items-center justify-center"
                style="width: 100%; height: 600px;"
              >
                <div
                  class="text-center"
                >
                  <div
                    class="inline-block animate-spin rounded-full h-8 w-8 border-b-2 border-green-600"
                  />
                  <p
                    class="mt-2 text-gray-600"
                  >
                    Loading map...
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="lg:col-span-1"
        >
          <div
            data-testid="mock-campsite-list"
          >
            <div>
              高尾山キャンプ場
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
